### DISCUSSION TITLE
AbstractGalerkin.jl

### DISCUSSION THREAD

**Discussion by DanielBoigk** [GalerkinToolkit/GalerkinToolkit.jl#260] [Ideas]
Background: I needed to use a FEM library for a reconstruction task called Electrical Impedance Tomography, which involves inverse problems on PDE's, Manifold optimization, Machine Learning.
- I tried Gridap: But ran into problems with the DSL, the speed I got out of it. The integration with Lux.jl which I wanted.
- I switched to Ferrite.jl

I have some questions:
- Would you consider doing a Package AbstractGalerkin.jl which basically just gives Abstract Types and methods that just define operations one might do on a Hilbertspace/SobolevSpace? Like projection to boundary. Extraction of points. scalar products, norms, vectorspace operations, (bi)linear forms, differentiation, integration, domain handling, etc.  Export of functions & Operators. And define a DSL there first and then create the connection to FEM, Spectral methods (FFT, Chebyshev), Wavelets, etc. or whatever some Julia library (like Ferrite, FFTW, ApproxFun, ...) is capable of ?

- Would you consider implementing methods that allow mixing of methods, like:
	- My EIT reconstruction would prefer spectral methods that act globally over the whole domain for an initial reconstruction.
	- Then I would like FEM to refine the points for the residual where higher resolution is possible.

- Let's say someone comes with a convolutional network in Lux.jl which in a first step samples over the grid via a filter/convolution and then predicts some output. Would it also be somewhat feasible to run such a NN in a training loop with GalerkinToolkit.jl?

- Would you make it so that each operation can be exported as a Matrix, Function, Tensor,etc? Or that it would be connectable to something like ModelingToolkit.jl? ODE solvers, DAE solvers?

It would be nice if you people would lay out your plans, and maybe If my time/job allows I could even contribute.

**Comment by fverdugo** [comment]
Hi @DanielBoigk 

> Would you make it so that each operation can be exported as a Matrix, Function, Tensor,etc? Or that it would be connectable to something like ModelingToolkit.jl? ODE solvers, DAE solvers?

GalerkinToolkit provides tools to assemble multi-linear forms on structured and unstructured meshes, with a DSL (similar to Gridap) or iterating over quantities at integration points (similar to Ferrite).

It also provides ways of assembling some problem types that can be solved with the SciML ecosystem.

See the intro of the Manual for more info:

https://galerkintoolkit.github.io/GalerkinToolkit.jl/dev/manual/introduction/

> Would it also be somewhat feasible to run such a NN in a training loop with GalerkinToolkit.jl?

I am not sure if I understand your use case.

**Reply by DanielBoigk** [reply]
Let's say I wanna do adversarial training, where the Regularizer learns as an actor and the FEM solver acts as the critic. Like This: https://arxiv.org/pdf/1805.11572

So I have some reconstruction loop for my inverse problem. It solves a Finite Element problem in each step and updates some target variable. While this is running the regularizer will both do inference and train as an online step instead of on a fixed dataset.

My minimization functional looks something like this:
`J(σ,𝐟,𝐠) = β*R(σ) +  ∑Jᵢ(σ,fᵢ,gᵢ) `
With R being a Neural Network and Jᵢ being the contributions of the FEM solver. and |σ-σ_true| my loss.

I struggled doing that in Gridap.

What you guys are about to build looks awesome!

**Comment by fverdugo** [comment]
@DanielBoigk can you tell which particular operations you need from the FEM solver? I.e., who are the inputs σ,fᵢ,gᵢ and how are they represented in the code, and how do you compute J_i from these inputs.

**Reply by DanielBoigk** [reply]
Basic: Description: There is a medium (i.e. domain) with a (complex) valued electrical conductivity σ (capacitance and inductance being the complex part). On some points at the boundary are electrodes, via these electrodes electrical current (g) is inserted (as a neumann boundary) and voltages (f) are measured (as dirichlet boundary). (note that f(t) and g(t) are actually signals that vary over time. And g(t) can be chosen by the reconstruction algorithm.) 
Note that the domain shape and conductivity can be a moving target: People breathe if one does lung tomography, arteries are moving blood, etc.  
In neumann boundary conditions the equation is:

$$\nabla \cdot(\sigma\nabla u(x,t)) = 0\quad\quad \forall x\in \Omega$$

and:

$$\sigma \frac{\partial u(x,t)}{\partial n} = g(x,t) \quad\quad \forall x\in \partial\Omega$$

Then there is a boundary projection proj_∂Ω (Which can be/is different from the neumann boundary )
Then the error is 
$$d( proj_{\partial \Omega}(u(t)) , f(t) )$$
with d some metric (like L^2 metric on the boundary)
Goal is to reconstruct σ from the boundary measurements f(t),g(t).
J_i ends up being something like

$$J_i(\sigma,f_i,g_i) = d(\text{proj}_{\partial\Omega}(L_\sigma^{-1}g_i),f_i)$$

Here $L_\sigma =$∫(σ * ∇(u)⋅∇(v))dΩ is assembled each time we change $\sigma$

I'd say there are boundary pairs $(f_1,g_1), ..., (f_n,g_n)$ because only a finite amount of such boundary snapshots can be calculated at the same time.

Machine Learning comes to play:
- for the regularizer 
- for the metric d (there metrics which are more aware of our prior )
- For the Reinforcement Learning methods we are considering that select the signal for the input current.

The problem is just really ill-conditioned and nonlinear, already in the reconstruction process one needs to make lot's of data driven assumptions else it's gonna converge nowhere near the correct solution.

**Reply by fverdugo** [reply]
Hi, I would say that this is relatively standard, unless you want to compute sensitivities wrt mesh position.

Not that you will provably need to build and solve an adjoint problem, using building blocks provided by GT or Gridap.
